### PR TITLE
DiegoProjectSearchDev

### PR DESCRIPTION
**Pull request by DIEGOTEC18**
Create a function that queries the project datastore by creating single and composite filters depending on the user search by tags, implement FuzzySearch (implements fuzzywuzzy by @xdrop) using Levenshtein distance to determine how similar is a project title to the free text of the user filtering the already queried projects. Modify the doGet function in the servlet to create SearchProject Objects from the resultant projects and send them in a JSON string in the response.

**IMPORTANT:** This pull request also includes @Manniec accidental merge.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 1]
```diff
@@ -8,69 +8,223 @@
...
+        //Create a JSONString from the ArrayList:
+        String json = toJsonString(projects);
+
+        //Write the JSON to the response:
+        response.setContentType("application/json; charset=utf-8");
+        response.getWriter().println(json);
+
+    }
+
+    //This post method is only to put dummy entities in the datastore for test purposes:
```
Comment (marcelpuyat) [inline src/main/java/com/google/sps/servlets/SearchServlet.java:91]: Nice test method for making it easier to verify functionality!

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 2]
```diff
@@ -8,69 +8,223 @@
...
 import com.google.gson.Gson;
+import java.sql.Timestamp;
 
 @WebServlet("/search")
 public class SearchServlet extends HttpServlet {
 
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
+        //Get the datastore instance:
```
Comment (marcelpuyat) [inline src/main/java/com/google/sps/servlets/SearchServlet.java:35]: The comments you placed are really helpful for making it easier to understand the flow.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 3]
```diff
@@ -223,8 +203,46 @@ private PreparedQuery queryProjectDatastore(HttpServletRequest request, Datastor
...
-            return results;
+            return finalResults;
+
+    }
+
+    //Filters the already filtered entities from the datastore query by it's title similarity to the searchBarText:
+    private Iterable<Entity> fuzzySearch(String searchText, Iterable<Entity> entities){
+
+        //0 is completely equal. 100 is equal:
+        int threshold = 30;
```
Comment (marcelpuyat) [inline src/main/java/com/google/sps/servlets/SearchServlet.java:227]: usually it's best practice to place all constants as final static variables of the class. i.e.

private static final fuzzySearchThreshold = 30;

at the top of the class.

This way it's easy to find and modify these values later on if needed.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 4]
```diff
@@ -223,8 +203,46 @@ private PreparedQuery queryProjectDatastore(HttpServletRequest request, Datastor
...
+
+        //ArrayList of projects that matched the free text search:
+        ArrayList<Entity> matches = new ArrayList<Entity>();
+
+        //Do a FuzzySearch between the searchBarText and every entity's title:
+        for(Entity entity : entities){
+
+            String currentTitle = (String) entity.getProperty("title");
+
+            if(FuzzySearch.ratio(searchText, currentTitle) > threshold){
```
Comment (marcelpuyat) [inline src/main/java/com/google/sps/servlets/SearchServlet.java:237]: Nicely done!

**Review by marcelpuyat — APPROVED** [review]
Great work Diego! Just a few minor comments.

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 2 above)
Comment (DIEGOTEC18) [reply src/main/java/com/google/sps/servlets/SearchServlet.java:35]: Thanks!

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 3 above)
Comment (DIEGOTEC18) [reply src/main/java/com/google/sps/servlets/SearchServlet.java:227]: Noted! It will be changed on my next commit.

Comment (DIEGOTEC18) [reply src/main/java/com/google/sps/servlets/SearchServlet.java:227]: Also, my comment is wrong there. It should be 

> 0 is completely different. 100 is equal:

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 4 above)
Comment (DIEGOTEC18) [reply src/main/java/com/google/sps/servlets/SearchServlet.java:237]: Thanks!

**Pull request merged**